xdem/spatial_tools.py

```diff
@@ -398,6 +408,149 @@ def hillshade(dem: Union[np.ndarray, np.ma.masked_array], resolution: Union[floa
+        raise ValueError('Attribute linear_pkg must be one of "scipy" or "sklearn".')
+
+    # select sklearn estimator
+    dict_estimators = {'Linear': LinearRegression(), 'Theil-Sen':TheilSenRegressor(random_state=random_state)
+        , 'RANSAC': RANSACRegressor(random_state=random_state), 'Huber': HuberRegressor()}
+    est = dict_estimators[estimator]
+
+    # TODO: this should be a function outside (waiting for Amaury's PR)
+    # remove NaNs
+    mykeep = np.logical_and.reduce((np.isfinite(y), np.isfinite(x)))
+    x = x[mykeep]
+    y = y[mykeep]
+
+    # subsample
+    index_sub = subsample_raster(x, 25000, return_indices=True)
+    x = x[index_sub]
+    y = y[index_sub]
+
+    # initialize cost function and output coefficients
+    mycost = np.empty(max_order)
+    coeffs = np.zeros((max_order, max_order + 1))
+    # loop on polynomial degrees
+    for deg in np.arange(1, max_order + 1):
+        # if method is linear, and package is scipy
+        if estimator == 'Linear' and linear_pkg == 'scipy':
+            # define the residual function to optimize
+            def fitfun_polynomial(xx, params):
+                return sum([p * (xx ** i) for i, p in enumerate(params)])
+            def errfun(p, xx, yy):
+                return fitfun_polynomial(xx, p) - yy
+            p0 = np.polyfit(x, y, deg)
+            myresults = scipy.optimize.least_squares(errfun, p0, args=(x, y), **kwargs)
+            if verbose:
+                print("Initial Parameters: ", p0)
+                print("Polynomial degree - ", deg, " --> Status: ", myresults.success, " - ", myresults.status)
+                print(myresults.message)
+                print("Lowest cost:", myresults.cost)
+                print("Parameters:", myresults.x)
+            mycost[deg - 1] = myresults.cost
+            coeffs[deg - 1, 0:myresults.x.size] = myresults.x
+        # otherwise, it's from sklearn
+        else:
+            if not _has_sklearn:
+                raise ValueError("Optional dependency needed. Install 'scikit-learn'")
+            p = PolynomialFeatures(degree=deg)
+            model = make_pipeline(p, est)
+            model.fit(x.reshape(-1,1), y)
+            mad = cost_func(model.predict(x.reshape(-1,1)), y)
+            mycost[deg - 1] = mad
+            # get polynomial estimated with the estimator
+            if estimator in ['Linear','Theil-Sen','Huber']:
+                c = est.coef_
+            # for some reason RANSAC doesn't store coef at the same place
+            elif estimator == 'RANSAC':
+                c = est.estimator_.coef_
+            coeffs[deg - 1, 0:deg+1] = c
+
+    # selecting the minimum (not robust)
+    # fidx = mycost.argmin()
+    # choosing the best polynomial with a margin of improvement on the cost
+    fidx = choice_best_polynomial(cost=mycost, margin_improvement=margin_improvement, verbose=verbose)
+
+    return np.trim_zeros(coeffs[fidx], 'b'), fidx+1
```

Comment:
Why is it `fixd + 1` and not `fixd`? Also, what does the "fix" part mean of the variable name?

Reply:
Added a comment for clarification, and renamed into `final_index`